ko/lessons/advanced/behaviours.md

```diff
@@ -0,0 +1,118 @@
+---
+layout: page
+title: 비헤이비어
+category: advanced
+order: 10
+lang: ko
+---
+
+이전 강좌에서 타입스펙에 관해 배웠습니다. 여기서는 그 사양을 구현하기 위해 모듈을 요청하는 방법을 배우겠습니다. Elixir에서 이 기능은 비헤이비어라는 이름으로 불립니다.
+
+{% include toc.html %}
+
+## 용도
+
+가끔 공개 API를 공유하는 모듈을 만들어야 할 때가 있습니다. Elixir에서 이런 경우의 해결책은 비헤이비어입니다. 비헤이비어는 두 주요 역할을 수행합니다.
+
++ 구현해야만 하는 함수의 목록을 정의
++ 그 함수가 실제로 구현되었는지 확인
+
+Elixir는 GenServer를 비롯해 비헤이비어를 여럿 가지고 있습니다만, 이 강좌에서는 그런 것을 사용하기 보단 직접 만드는것에 집중해 보겠습니다.
+
+## 비헤이비어 정의하기
+
+비헤이비어를 좀 더 잘 이해하기 위해 워커 모듈을 위한 비헤이비어를 구현해 보겠습니다. 이 워커는 `init/1`, `perform/2` 두 함수가 구현되어 있어야 합니다.
+
+이를 충족하기 위해, `@spec`과 비슷한 `@callback` 디렉티브를 사용하겠습니다. 이 디렉티브는 메서드가 반드시 __필요__합니다. 매크로는 `@macrocallback`를 사용해서 할 수 있습니다. 워커를 위해 `init/1`, `perform/2` 매서드를 정의해 봅니다.
+
+```elixir
+defmodule Example.Worker do
+  @callback init(state :: term) :: {:ok, new_state :: term} | {:error, reason :: term}
+  @callback perform(args :: term, state :: term) :: {:ok, result :: term, new_state :: term} | {:error, reason :: term, new_state :: term}
+end
+```
+
+여기에 정의된 `init/1`는 어떤 값을 받아 튜플 `{:ok, state}`나 `{:error, reason}`를 반환합니다. 이는 초기화에서 꽤 일반적인 패턴입니다. `perform/2` 메서드는 어떤 인자를 받아 워커의 상태를 초기화합니다. `perform/2`는 `{:ok, result, state}`나 `{:error, reason, state}`를 반환하길 기대하고 이는 GenServer와 비슷합니다.
+
+## 비헤이비어 사용하기
+
+이제 비헤이비어를 정의했으니 공개 API를 공유하는 다양안 모듈에서 이를 사용할 수 있습니다. 비헤이비어를 모듈에 추가하는 것은 `@behaviour` 속성을 사용하면 쉽습니다.
+
+새로운 비헤이비어를 사용해 원격에서 파일을 다운로드해 로컬에 저장하는 모듈 테스크를 만들어 봅시다.
+
+```elixir
+defmodule Example.Downloader do
+  @behaviour Example.Worker
+
+  def init(opts), do: {:ok, opts}
+
+  def perform(url, opts) do
+    url
+    |> HTTPoison.get!
+    |> Map.fetch(:body)
+    |> write_file(opts[:path])
+    |> respond(opts)
+  end
+
+  defp write_file(:error, _), do: {:error, :missing_body}
+  defp write_file({:ok, contents}, path) do
+    path
+    |> Path.expand
+    |> File.write(contents)
+  end
+
+  defp respond(:ok, opts), do: {:ok, opts[:path], opts}
+  defp respond({:error, reason}, opts), do: {:error, reason, opts}
+end
+```
+
+아니면 파일의 목록을 압축하는 워커는 어떨까요? 그것도 가능합니다.
```

Comment:
> 아니면 파일의 목록을 받아 압축하는 워커는 어떨까요?

That really sounds like compressing the text (which contains a list of file) :joy: